quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -123,12 +127,24 @@ class
   nextState :: Typeable a => state -> Action state a -> Var a -> state
   nextState s _ _ = s
 
+  -- | Transition function for negative actions. Note that most negative testing applications
+  -- should not require an implementation of this function!
+  negativeNextState :: Typeable a => state -> Action state a -> state
+  negativeNextState s _ = s
+
   -- | Precondition for filtering generated `Action`.
   -- This function is applied before the action is performed, it is useful to refine generators that
   -- can produce more values than are useful.
   precondition :: state -> Action state a -> Bool
   precondition _ _ = True
 
+  -- | Precondition for filtering an `Action` that can meaningfully run but is supposed to fail.
+  -- An action will run as a negative action if it's precondition fails. This action should have
+  -- _no effect_ on the model state. This may not be desierable in all situations - in which case
+  -- one can override this semantics for book-keeping in `negativeNextState`.
+  negativePrecondition :: state -> Action state a -> Bool
```

Comment:
Suggestion:  `negativePrecondition` does not really convey the intent of this function, perhaps something like `failIfRun` or `expectedToFail` or something similar would perhaps help clarify it.

Reply:
I replaced it with `validFailingAction`.